nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java

```diff
@@ -107,110 +173,212 @@ public NullnessHint onDataflowVisitMethodInvocation(
+            // clause for the purpose of deciding the nullness of the other arguments.
             continue;
-          } else if (valueConstraint.equals("null") || valueConstraint.equals("!null")) {
-            if (argIdx != -1) {
-              // More than one argument involved in the antecedent, ignore this rule
-              supported = false;
-              break;
-            }
-            argIdx = i;
-            argAntecedentNullness =
-                valueConstraint.equals("null") ? Nullness.NULLABLE : Nullness.NONNULL;
-          } else {
-            String errorMessage =
-                "Invalid @Contract annotation detected for method "
-                    + callee
-                    + ". It contains the following uparseable clause: "
-                    + clause
-                    + " (unknown value constraint: "
-                    + valueConstraint
-                    + ", see https://www.jetbrains.com/help/idea/contract-annotations.html).";
-            state.reportMatch(
-                analysis
-                    .getErrorBuilder()
-                    .createErrorDescription(
-                        new ErrorMessage(
-                            ErrorMessage.MessageTypes.ANNOTATION_VALUE_INVALID, errorMessage),
-                        tree,
-                        analysis.buildDescription(tree),
-                        state,
-                        null));
+          }
+          if (arg != null) {
+            // More than one argument involved in the antecedent, ignore this rule
             supported = false;
             break;
           }
-        }
-        if (!supported) {
-          // Too many arguments involved, or unsupported @Contract features. On to next clause in
-          // the
-          // contract expression
+          arg = nullTestTarget;
+          argAntecedentNullness = antecedentNullness;
+        } else if (valueConstraint.equals("!null")
+            && inputs.valueOfSubNode(node.getArgument(i)).equals(Nullness.NONNULL)) {
+          // We already know this argument can't be null, so we can treat it as not part of the
+          // clause
+          // for the purpose of deciding the non-nullness of the other arguments.
           continue;
-        }
-        if (argIdx == -1) {
-          // The antecedent is unconditionally true. Check for the ... -> !null case and set the
-          // return nullness accordingly
-          if (consequent.equals("!null")) {
-            return NullnessHint.FORCE_NONNULL;
+        } else if (valueConstraint.equals("null") || valueConstraint.equals("!null")) {
+          if (arg != null) {
+            // More than one argument involved in the antecedent, ignore this rule
+            supported = false;
+            break;
           }
-          continue;
+          arg = node.getArgument(i);
+          argAntecedentNullness = valueConstraint.equals("null") ? Nullness.NULL : Nullness.NONNULL;
```

Comment:
I think this comment also helps me understand your comment in https://github.com/uber/NullAway/pull/669#discussion_r989495957

I believe there are two things going on: First, we're not confident about trusting the value of `Nullness.NULL` as an input to checks, which I'll pull out to another issue or PR for discussion separately. Second, the reliance on `Nullness.NULL` with regard to the `argAntecedentNullness` variable.

I think we're in agreement about trimming out the `NULL` pieces as inputs to checks in this pull request, however I don't believe `argAntecedentNullness` `Nullness` values themselves are connected to dataflow analysis at all beyond reusing the same data-type -- this was confusing for me when I began investigating this source file because the argAntecedentNullness represents a constraint, which is either `NULL` or `NONNULL`, where we can't reason about cases in which the arg is neither, if that makes sense. To summarize, we could replace all the assignments and checks of `Nullness.NULL` specifically on `argAntecedentNullness` with `Nullness.NULLABLE` (or even Nullness.BOTTOM, for that matter), and the results would not be impacted because the value of that variable is confined to this method.